File: scripts/wechat-pay-probe.mjs
#!/usr/bin/env node
/**
 * 探测微信支付 v3 签名（SIGN_ERROR 时用）
 * 用法: node scripts/wechat-pay-probe.mjs
 */
import { createPrivateKey, randomBytes, sign } from 'crypto';
import { existsSync, readFileSync } from 'fs';
import https from 'https';
import { dirname, resolve } from 'path';
import { fileURLToPath } from 'url';

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const ENV_FILE = [resolve(ROOT, '.env'), resolve(ROOT, 'server/.env')].find(existsSync);

function loadEnv(file) {
  const env = {};
  for (const line of readFileSync(file, 'utf8').split('\n')) {
    const t = line.trim();
    if (!t || t.startsWith('#')) continue;
    const i = t.indexOf('=');
    if (i <= 0) continue;
    let v = t.slice(i + 1).trim();
    if ((v.startsWith('"') && v.endsWith('"')) || (v.startsWith("'") && v.endsWith("'"))) {
      v = v.slice(1, -1);
    }
    env[t.slice(0, i).trim()] = v;
  }
  return env;
}

function loadKey(env) {
  let pem = (env.WECHAT_PRIVATE_KEY || '').trim().replace(/\\n/g, '\n');
  if (!pem && env.WECHAT_PRIVATE_KEY_PATH) {
    pem = readFileSync(env.WECHAT_PRIVATE_KEY_PATH.trim(), 'utf8').trim();
  }
  if (!pem) throw new Error('未配置 WECHAT_PRIVATE_KEY 或 WECHAT_PRIVATE_KEY_PATH');
  return pem;
}

function buildAuth(method, urlPath, body, env, keyPem) {
  const mchId = env.WECHAT_MCH_ID.trim();
  const serial = env.WECHAT_SERIAL_NO.trim();
  const ts = Math.floor(Date.now() / 1000).toString();
  const nonce = randomBytes(16).toString('hex');
  const message = `${method}\n${urlPath}\n${ts}\n${nonce}\n${body}\n`;
  const key = createPrivateKey(keyPem);
  const sig = sign('RSA-SHA256', Buffer.from(message, 'utf8'), key).toString('base64');
  return (
    `WECHATPAY2-SHA256-RSA2048 mchid="${mchId}",` +
    `nonce_str="${nonce}",signature="${sig}",` +
    `timestamp="${ts}",serial_no="${serial}"`
  );
}

function request(method, urlPath, bodyStr, env, keyPem) {
  return new Promise((resolve, reject) => {
    const auth = buildAuth(method, urlPath, bodyStr, env, keyPem);
    const bodyBuf = bodyStr ? Buffer.from(bodyStr, 'utf8') : null;
    const req = https.request(
      {
        hostname: 'api.mch.weixin.qq.com',
        path: urlPath,
        method,
        headers: {
          Accept: 'application/json',
          Authorization: auth,
          'User-Agent': 'wx-group-probe',
          ...(bodyBuf
            ? {
                'Content-Type': 'application/json',
                'Content-Length': bodyBuf.length,
              }
            : {}),
        },
      },
      (res) => {
        let text = '';
        res.on('data', (c) => (text += c));
        res.on('end', () => resolve({ status: res.statusCode, text }));
      },
    );
    req.on('error', reject);
    if (bodyBuf) req.write(bodyBuf);
    req.end();
  });
}

if (!ENV_FILE) {
  console.error('未找到 .env');
  process.exit(1);
}

const env = loadEnv(ENV_FILE);
const keyPem = loadKey(env);

console.log('========== 配置 ==========');
console.log('MCH_ID', env.WECHAT_MCH_ID);
console.log('APP_ID', env.WECHAT_APP_ID);
console.log('SERIAL', env.WECHAT_SERIAL_NO);
console.log('KEY', env.WECHAT_PRIVATE_KEY_PATH || '(inline)');
console.log('');

console.log('========== 1. GET /v3/certificates（仅测签名，无 body）==========');
const r1 = await request('GET', '/v3/certificates', '', env, keyPem);
console.log('HTTP', r1.status, r1.text.slice(0, 300));
if (r1.text.includes('SIGN_ERROR')) {
  console.log('\n✗ 连 GET 都 SIGN_ERROR → 商户号/序列号/私钥 与微信商户平台不一致');
  console.log('  请登录 pay.weixin.qq.com → API安全 → 确认「商户API证书」序列号与 .env 完全一致');
  console.log('  并与小程序服务器对比 apiclient_key.pem 的 md5');
  process.exit(1);
}
console.log('✓ GET 签名通过\n');

console.log('========== 2. POST jsapi（模拟下单，openid 占位）==========');
const jsapiBody = JSON.stringify({
  appid: env.WECHAT_APP_ID.trim(),
  mchid: env.WECHAT_MCH_ID.trim(),
  description: 'probe-test',
  out_trade_no: `PROBE${Date.now()}`,
  notify_url: (env.WECHAT_NOTIFY_URL || 'https://jiaoyou.yikuaikaixin.cn/api/v1/payments/wechat/notify').trim(),
  amount: { total: 1, currency: 'CNY' },
  payer: { openid: 'oUpF8uMuAJO_M2pxb1Q9zNjWeS6o' },
});
const r2 = await request('POST', '/v3/pay/transactions/jsapi', jsapiBody, env, keyPem);
console.log('HTTP', r2.status, r2.text.slice(0, 400));
if (r2.text.includes('SIGN_ERROR')) {
  console.log('\n✗ POST jsapi SIGN_ERROR（GET 已通过，少见）');
} else if (r2.text.includes('openid') || r2.text.includes('APPID')) {
  console.log('\n✓ POST 签名通过（openid/appid 报错属正常，说明签名 OK）');
} else {
  console.log('\n响应见上');
}

File: scripts/wechat-pay-probe.py
#!/usr/bin/env python3
"""与 wander_meet wechat_pay_v3.py 相同签名逻辑，用于对比 Node probe。"""
from __future__ import annotations

import base64
import json
import os
import secrets
import sys
import time
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
ENV_FILE = ROOT / ".env"
if not ENV_FILE.exists():
    ENV_FILE = ROOT / "server" / ".env"


def load_env() -> dict[str, str]:
    env: dict[str, str] = {}
    for line in ENV_FILE.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        k, v = line.split("=", 1)
        env[k.strip()] = v.strip().strip('"').strip("'")
    return env


def load_key(env: dict[str, str]) -> str:
    pem = (env.get("WECHAT_PRIVATE_KEY") or "").strip().replace("\\n", "\n")
    path = (env.get("WECHAT_PRIVATE_KEY_PATH") or "").strip()
    if not pem and path:
        pem = Path(path).read_text(encoding="utf-8")
    if not pem:
        raise SystemExit("未配置 WECHAT_PRIVATE_KEY 或 WECHAT_PRIVATE_KEY_PATH")
    return pem.strip()


def sign_message(message: str, pem: str) -> str:
    from cryptography.hazmat.primitives import hashes, serialization
    from cryptography.hazmat.primitives.asymmetric import padding

    key = serialization.load_pem_private_key(pem.encode("utf-8"), password=None)
    sig = key.sign(message.encode("utf-8"), padding.PKCS1v15(), hashes.SHA256())
    return base64.b64encode(sig).decode("ascii")


def auth(method: str, url_path: str, body: str, env: dict[str, str], pem: str) -> str:
    mch_id = env["WECHAT_MCH_ID"].strip()
    serial = env["WECHAT_SERIAL_NO"].strip()
    ts = str(int(time.time()))
    nonce = secrets.token_hex(16)
    message = f"{method}\n{url_path}\n{ts}\n{nonce}\n{body}\n"
    signature = sign_message(message, pem)
    return (
        f'WECHATPAY2-SHA256-RSA2048 mchid="{mch_id}",'
        f'nonce_str="{nonce}",signature="{signature}",'
        f'timestamp="{ts}",serial_no="{serial}"'
    )


def request(method: str, url_path: str, body: str, env: dict[str, str], pem: str) -> tuple[int, str]:
    url = f"https://api.mch.weixin.qq.com{url_path}"
    headers = {
        "Authorization": auth(method, url_path, body, env, pem),
        "Accept": "application/json",
        "User-Agent": "wx-group-python-probe",
    }
    data = body.encode("utf-8") if body else None
    if data:
        headers["Content-Type"] = "application/json"
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=20) as resp:
            return resp.status, resp.read().decode("utf-8")
    except urllib.error.HTTPError as e:
        return e.code, e.read().decode("utf-8")


def main() -> None:
    if not ENV_FILE.exists():
        raise SystemExit(f"未找到 {ENV_FILE}")

    env = load_env()
    pem = load_key(env)

    print("========== Python probe (wander_meet 同款) ==========")
    print("ENV", ENV_FILE)
    print("MCH_ID", env.get("WECHAT_MCH_ID"))
    print("SERIAL", env.get("WECHAT_SERIAL_NO"))
    print()

    status, text = request("GET", "/v3/certificates", "", env, pem)
    print("GET /v3/certificates", status, text[:400])
    if "SIGN_ERROR" in text:
        print("\n✗ Python 也 SIGN_ERROR → 100% 是商户平台证书/私钥/序列号问题，不是 Node 代码")
        print("  1. pay.weixin.qq.com → API安全 → 查看「商户API证书」当前序列号")
        print("  2. 重新下载 API 证书 zip，替换 apiclient_key.pem 并更新 WECHAT_SERIAL_NO")
        print("  3. 从小程序服务器复制 WECHAT_PAY_PRIVATE_KEY 或 apiclient_key.pem（md5 必须一致）")
        sys.exit(1)
    print("\n✓ Python GET 签名通过")


if __name__ == "__main__":
    main()
